A83-xds-gcp-authn-filter.md

```diff
@@ -0,0 +1,218 @@
+In this message, the `url` field must be non-empty; if empty, the
+resource will be NACKed.  The parsed representation of this message can
+be a simple string.
+
+Note that in Envoy, the metadata keys must exactly match the legacy
+filter name (e.g., "envoy.filters.http.gcp_authn").  However, as per
+envoyproxy/envoy#34251, it is desirable to instead use the HTTP filter
+instance name from the [`HttpFilter.name`
+field](https://github.com/envoyproxy/envoy/blob/7436690884f70b5550b6953988d05818bae3d087/api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#L1149).
+We will implement that behavior in gRPC.
+
+### xDS ConfigSelector Behavior
+
+As per [gRFC A60][A60], we currently pass the selected cluster name via
+a call attribute for access in filters.  We will now also pass the CDS
+resource for the cluster in a separate call attribute, so that the GCP
+Authentication filter can access the cluster metadata for the selected
+cluster.  Note that this will require implementing [A74], so that the
+CDS resource is available at xDS routing time.
+
+### Filter Behavior
+
+The filter will maintain a cache of
+GcpServiceAccountIdentityCallCredentials instances, one for each
+audience, along with a last-used timestamp for each one.  The maximum
+number of entries in the cache is bounded by the config field
+`cache_config.cache_size`; if the cache exceeds that size, then the
+entry with the oldest last-used timestamp will be removed.
+
+When the filter processes the RPC's initial metadata, it will first
+attempt to determine the audience for the request by looking at the
+cluster metadata key corresponding to the filter instance name, which
+must be of type `extensions.filters.http.gcp_authn.v3.Audience`.  If the
+cluster has no such metadata key, the filter is a no-op.  Otherwise, the
+audience is the value of the `url` field in the `Audience` proto.
+
+The filter will then check to see if it already has a cached
+GcpServiceAccountIdentityCallCredentials instance for the specified
+audience.  If it does not, it will create a new instance, adding it to
+its cache, removing the least recently used entry from the cache if the
+cache is already at its max size.  It will then attach that
+GcpServiceAccountIdentityCallCredentials instance to the RPC.
+
+Note that implementations must ensure that the token is not added to
+RPCs sent on insecure connections.  However, the GCP Authentication
+filter will run before load balancing has chosen a connection, so the
+filter cannot directly add the token to the RPC.  Instead, it must add
+the call credential to the RPC, and the call credential will do the work
+of adding the token to the RPC later, after load balancing has chosen a
+connection.
+
+### Temporary environment variable protection
+
+Support for the GCP Authentication filter in LDS will be guarded by the
```

Comment:
It should also protect extensions.filters.http.gcp_authn.v3.Audience from being in the metadata registry.